File: kapp-mvp/src/main/java/com/taishi/kapp_mvp/presenter/RepositoryPresenter.java
package com.taishi.kapp_mvp.presenter;

import android.util.Log;

import com.taishi.kapp_mvp.ArchiApplication;
import com.taishi.kapp_mvp.GithubService;
import com.taishi.kapp_mvp.User;
import com.taishi.kapp_mvp.view.RepositoryMvpView;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Action1;


public class RepositoryPresenter implements Presenter<RepositoryMvpView> {

    private static final String TAG = "RepositoryPresenter";

    private RepositoryMvpView repositoryMvpView;
    private Subscription subscription;

    @Override
    public void attachView(RepositoryMvpView view) {
        this.repositoryMvpView = view;
    }

    @Override
    public void detachView() {
        this.repositoryMvpView = null;
        if (subscription != null) subscription.unsubscribe();
    }

    public void loadOwner(String userUrl) {
        ArchiApplication application = ArchiApplication.get(repositoryMvpView.getContext());
        GithubService githubService = application.getGithubService();
        subscription = githubService.userFromUrl(userUrl)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(application.defaultSubscribeScheduler())
                .subscribe(new Action1<User>() {
                    @Override
                    public void call(User user) {
                        Log.i(TAG, "Full user data loaded " + user);
                        repositoryMvpView.showOwner(user);
                    }
                });
    }
}

File: kapp-mvp/src/main/java/com/taishi/kapp_mvp/presenter/RepositoryPresenter.kt
package com.taishi.kapp_mvp.presenter

import android.util.Log
import com.taishi.kapp_mvp.ArchiApplication
import com.taishi.kapp_mvp.view.RepositoryMvpView
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers


class RepositoryPresenter : Presenter<RepositoryMvpView> {

    private var repositoryMvpView: RepositoryMvpView? = null
    private var subscription: Subscription? = null

    override fun attachView(view: RepositoryMvpView) {
        this.repositoryMvpView = view
    }

    override fun detachView() {
        this.repositoryMvpView = null
        if (subscription != null) subscription!!.unsubscribe()
    }

    fun loadOwner(userUrl: String) {
        val application = ArchiApplication[repositoryMvpView!!.getContext()]
        val githubService = application.githubService
        subscription = githubService!!.userFromUrl(userUrl)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeOn(application.defaultSubscribeScheduler())
                .subscribe { user ->
                    Log.i(TAG, "Full user data loaded " + user)
                    repositoryMvpView!!.showOwner(user)
                }
    }

    companion object {

        private val TAG = "RepositoryPresenter"
    }
}
